Add bookings

JustAlexB opened the pull request.

JustAlexB committed:
refactor: первый релиз shareit

JustAlexB committed:
fix: check-style-plugin formated

JustAlexB committed:
fix: equals and hasCode bug

JustAlexB committed:
fix: the bugs has been fixed

JustAlexB committed:
fix: check style

JustAlexB committed:
refactor: устранил замечания ко 2му релизу

JustAlexB committed:
feat: добавил бронивание и отзывы

JustAlexB committed:
fix: style check

JustAlexB committed:
fix: 'static' modifier deleted

JustAlexB committed:
fix: вернул строчку в настройки  on-profile

JustAlexB committed:
fix: use inMemory H2 DB

Dmitry-E commented:
![image](https://user-images.githubusercontent.com/45330610/228029132-95264c82-b126-4611-8321-dcaeda92d024.png)

ветка сделана не от main
перед тем как создать ветку add-bookings нужно было смержить предыдущую в main. Затем запулить main и после этого можно создавать add-bookings.

Dmitry-E commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,60 @@
 @RequestMapping(path = "/bookings")
+@Slf4j
 public class BookingController {
+    private final BookingService bookingService;
```
лучше сделать через интерфейс и реализацию

JustAlexB replied:
ok

Dmitry-E commented on src/main/java/ru/practicum/shareit/booking/service/BookingService.java:
```diff
@@ -0,0 +1,123 @@
+        if (fUser.isEmpty()) {
+            throw new NotFoundException("Пользователь не найден");
+        }
+        if (fItem == null) {
```
это условие всё время false

JustAlexB replied:
ok

Dmitry-E commented on src/main/java/ru/practicum/shareit/booking/service/BookingService.java:
```diff
@@ -0,0 +1,123 @@
+        }
+    }
+
+    public Booking addBooking(BookingDto bookingDto, Integer userID) {
```
метод обращается к нескольким репозиториям, поэтому нужно делать в одной транзакции.
добавляем `@Transactional`

JustAlexB replied:
ok

Dmitry-E commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -0,0 +1,35 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    Integer id;
+    @OneToOne(fetch = FetchType.EAGER)
```
лучше заменить `@OneToOne(fetch = FetchType.EAGER)` на 
	
```
@ManyToOne
@JoinColumn(name = "item_id")
```


и аналогично для booker

JustAlexB replied:
ok
Я неправильно выбрал связь или сама конструкция в чем-то лучше?

Dmitry-E replied:
тип связи правильней "многок к одному", т.к. одна запись item может быть в нескольких booking

JustAlexB replied:
Точно. сильно тормозил на прошлой неделе

Dmitry-E commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -0,0 +1,35 @@
+    User booker;
+    @Column(name = "status")
+    @Enumerated(EnumType.STRING)
+    BookingStatus status = BookingStatus.WAITING;
```
для обязательных проей можно проставить `nullable = false`

`@Column(name = "status", nullable = false)`

JustAlexB replied:
ok

Dmitry-E commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,60 @@
+    }
+
+    @PostMapping
+    public Booking addBooking(@Valid @RequestBody BookingDto booking,
```
возвращать нужно Dto

поля:
id, 
start
end
item (вложенный класс с id и name),
booker (вложенный класс с id и name), 
status

mapper нужно сделать вручную, т.е. создать new BookingDto и заполнить поля

JustAlexB replied:
ok

Dmitry-E commented on src/main/java/ru/practicum/shareit/booking/service/BookingService.java:
```diff
@@ -0,0 +1,123 @@
+        if (Arrays.stream(BookingState.values()).noneMatch(eState -> eState.name().equals(state))) {
+            throw new UnsupportedStatusException("UNSUPPORTED_STATUS");
+        }
+        List<Booking> allByState = bookingRepository.getAllByState(userId, state);
```
для разных статусов запросы идут с различными параметрами, т.е. делаем switch/case и под каждый статуст делаем соответствующий вызов.

аналогично и для getOwnerBookings

JustAlexB replied:
была такая идея, но показалась громозкой, решил не делать case
OK

JustAlexB replied:
Да, так точно лучше

Dmitry-E commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceDB.java:
```diff
@@ -0,0 +1,179 @@
+        this.commentStorage = commentStorage;
+    }
+
+    public Collection<ItemDto> getAll(Integer userID) {
```
метод используется только с userID, поэтому лучше оставить один вариант для userID != null

для id лучше использовать тип Long, т.к. данных может быть много

в контроллере можно прописать` long userID `и нам точно придёт не null

JustAlexB replied:
ok. Я все смотрел на этот Long в задании и теории и вообще не отразил ,что он сыграет со мной такую злую шутку. Думал - это перестраховка

Dmitry-E commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceDB.java:
```diff
@@ -0,0 +1,179 @@
+                .map(ItemDto::getId)
+                .collect(Collectors.toList());
+
+        List<BookingDetails> lastBookings = bookingStorage.getLastBookings(itemsId, LocalDateTime.now());
```
лучше получить все Booking для нужных itemsId за один запрос (сейчас в два) , и затем сгруппировать по items
(`Map<Item, List<Booking>>`)

JustAlexB replied:
Блин, у меня пока не получилось. Много времени потратил, решил оставить на второй круг. Извини.

JustAlexB replied:
ok

Dmitry-E commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceDB.java:
```diff
@@ -0,0 +1,179 @@
+            for (int i = 0; i < nextBookings.size(); i++) {
+                items.get(i).setNextBooking(nextBookings.get(i));
+            }
+        }
```
нужно ещё добавить Comments

JustAlexB replied:
Переписал там все! Добавил

Dmitry-E commented on src/main/java/ru/practicum/shareit/item/model/Comment.java:
```diff
@@ -0,0 +1,31 @@
+    @Column(nullable = false, length = 250)
+    String text;
+    @Column(name = "author_name")
+    String authorName;
```
```
   @Column(name = "author_name")
    String authorName;
```

лучше заменить на 
```
	@ManyToOne
	@JoinColumn(name = "author_id")
	private User author;
```

JustAlexB replied:
ok

JustAlexB committed:
fix: устранить большинство замечаний

1 пока не смог - много времени потратил на попытку из 2х обращени к репозиторию сдделать 1. Начал с него, но пототм понял ,что сильно отстаю и в этом релизе не успел его поправить

JustAlexB committed:
style: style check

JustAlexB committed:
refactor:  Suspicious comparison of Long

JustAlexB committed:
refactor: Item сделал интерфейс и реализацию

JustAlexB committed:
style: check

Dmitry-E commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,226 @@
+        List<Comment> allComments = commentStorage.findAllByItemsId(itemsId);
+
+        List<Booking> allLastBookings = bookingStorage
+                .findFirstByItem_IdInAndItem_Owner_IdAndStartIsBefore(
```
userId использовали для поиска items, и для поиска Bookings нам нужен только itemId из `items`
достаточно сделать один запрос к bookingStorage с фильтрацией по itemId и approve

```
Map<Item, List<Booking>> approvedBookings =
				bookingStorage.findApprovedForItems(items, Sort.by(DESC, "start"))
						.stream()
						.collect(groupingBy(Booking::getItem, toList()));
```
мы получим map из которого для каждой item можно быстро получить  `List<Booking>` и среди них найти нужную нам.

аналогично делаем для comments

JustAlexB replied:
я бы до такого в ближайшие недели точно не догуглил. Спасибо.

JustAlexB committed:
refactor: переписать на один вызов репозитория

JustAlexB committed:
style: check style

Dmitry-E commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,216 @@
+    private final UserRepository userStorage;
+    private final BookingRepository bookingStorage;
+    private final CommentRepository commentStorage;
+    private static final Sort SORT_DESC = Sort.by(DESC, "end");
```
неиспользуемые переменные лучше убрать

JustAlexB replied:
ok

Dmitry-E commented on src/main/java/ru/practicum/shareit/repository/BookingRepository.java:
```diff
@@ -0,0 +1,94 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+public interface BookingRepository extends JpaRepository<Booking, Long> {
```
в этом интерфейсе много неиспользуемого кода, надо почистить

JustAlexB replied:
да ,я что-то за грустью совсем забыл про него

JustAlexB replied:
странно, что check style проверяет всякую хрень и пропускает тонны закоментированного кода

Dmitry-E commented on src/main/java/ru/practicum/shareit/repository/BookingRepository.java:
```diff
@@ -0,0 +1,94 @@
+    List<Booking> getAllByStateRejected(@Param("userId") Long userId);
+
+    @Query("SELECT b FROM Booking b WHERE b.item.owner.id = :userId ORDER BY b.start DESC")
+    List<Booking> getAllForOwnerAll(@Param("userId") Long userId);
```
часть методов можно было сделать без явного указания запроса `@Query,` например  getAllForOwnerAll
`List<Booking> findByItem_Owner_Id(Long ownerId, Sort sort);`

JustAlexB replied:
сложно было, @Query легче тестировать в бобре

Dmitry-E commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,216 @@
+        return getItemsAnswers(items);
+    }
+
+    private Collection<ItemAnswerDto> getItemsAnswers(List<Item> items) {
```
делаем несколько запросов в разных транзакциях и в другом потоке могут добавить/изменить какие-то данные
т.к. тут только чтение, то можем добавить `@Transactional(readOnly = true)`

JustAlexB replied:
ок

Dmitry-E commented on src/main/java/ru/practicum/shareit/storage/DbUserStorage.java:
```diff
@@ -0,0 +1,64 @@
+
+    @Override
+    public User update(User user) {
+        if (user.getEmail() == null && user.getName() != null) {
```
если у пользователя добавятся новые поля, то тогда обновление потребует много запрсов.
лучше сначала получить из БД юзера
`userStorage.findById(user.getId())`, а затем дозаполнить и сохранить
  ```
      var oldUser = userStorage.findById(user.getId()).orElseThrow(...);
        if(...)
         newUser.setName();
       userRepository.save(newUser);
```

JustAlexB replied:
ok.

Dmitry-E commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,15 @@
+public class UserDto {
+    Long id;
+    String name;
+    @Email
```
при создании пользователя email обязательное, а при обновлении опциональное.
в таком случае можно сделать интерфейс "маркер" :

@```
Email(groups = {Update.class, Create.class})
@NotNull(groups = {Create.class})
 private String email;
```
в контроллере будем указывать `@Validated(Update.class) или @Validated(Create.class)`

сами интерфейсы пустые
```

public interface Create {
}
```

```
public interface Update {
}

```

JustAlexB replied:
ох, попробую

JustAlexB replied:
кажется, получилось

JustAlexB committed:
refactor: исправить замечания
